File: src/ports/outbound/investigation-synthesizer-runner.ts
import type { AlertContext } from "../../shared/types/alert-context";
import type { InvestigationResult } from "../../shared/types/investigation";
import type { LlmUsageSnapshot } from "../../shared/types/investigation-llm-telemetry";
import type { InvestigationProgressEventCallback } from "./investigation-progress-event";

const SYNTHESIZER_RUN_FAILED_ERROR_CODE = "SYNTHESIZER_RUN_FAILED";

export interface SynthesizerRunReport {
	reportText: string;
	usage: LlmUsageSnapshot;
}

export interface RunSynthesizerInput {
	result: InvestigationResult;
	alertContext: AlertContext;
	onProgressEvent: InvestigationProgressEventCallback;
}

interface SynthesizerRunFailedErrorInput {
	usage: LlmUsageSnapshot;
	cause: unknown;
}

export class SynthesizerRunFailedError extends Error {
	readonly code = SYNTHESIZER_RUN_FAILED_ERROR_CODE;
	readonly usage: LlmUsageSnapshot;
	override readonly cause: unknown;

	constructor(input: SynthesizerRunFailedErrorInput) {
		super("Synthesizer agent run failed");
		this.name = "SynthesizerRunFailedError";
		this.usage = input.usage;
		this.cause = input.cause;
	}
}

export interface InvestigationSynthesizerRunnerPort {
	run(input: RunSynthesizerInput): Promise<SynthesizerRunReport>;
}

export function isSynthesizerRunFailedError(
	error: unknown,
): error is SynthesizerRunFailedError {
	const candidate = Object(error) as Partial<SynthesizerRunFailedError>;
	return candidate.code === SYNTHESIZER_RUN_FAILED_ERROR_CODE;
}

File: rust/crates/shared/src/ports/outbound/investigation_synthesizer_runner.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::{
    errors::SynthesizerRunFailedError,
    types::{AlertContext, InvestigationResult, LlmUsageSnapshot},
};

use super::InvestigationProgressEventPort;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SynthesizerRunReport {
    pub report_text: String,
    pub usage: LlmUsageSnapshot,
}

pub struct RunSynthesizerInput {
    pub result: InvestigationResult,
    pub alert_context: AlertContext,
    pub on_progress_event: Arc<dyn InvestigationProgressEventPort>,
}

#[async_trait]
pub trait InvestigationSynthesizerRunnerPort: Send + Sync {
    async fn run(
        &self,
        input: RunSynthesizerInput,
    ) -> Result<SynthesizerRunReport, SynthesizerRunFailedError>;
}
